homeassistant/auth/providers/__init__.py

```diff
@@ -137,3 +143,116 @@ def async_create_credentials(self, data):
+        self._auth_provider = auth_provider
+        self._auth_module_id = None
+        self._auth_modules = []
+        self._auth_manager = auth_provider.hass.auth
+        self._user = None
+        self._username = None
+        self.created_at = dt_util.utcnow()
+
+    async def async_step_init(self, user_input=None):
+        """Handle the first step of login flow.
+
+        Return self.async_show_form(step_id='init') if user_input == None.
+        Return await self.async_finish(username) if login init step pass.
+        """
+        raise NotImplementedError
+
+    async def async_step_select_mfa_module(self, user_input=None):
+        """Handle the step of select mfa module."""
+        errors = {}
+
+        if user_input is not None:
+            auth_module = user_input.get('multi_factor_auth_module')
+            if auth_module in self._auth_modules:
+                self._auth_module_id = auth_module
+                return await self.async_step_mfa()
+            else:
+                errors['base'] = 'invalid_auth_module'
+
+        schema = OrderedDict()
+        schema['multi_factor_auth_module'] = vol.In(self._auth_modules)
+
+        return self.async_show_form(
+            step_id='select_mfa_module',
+            data_schema=vol.Schema(schema),
+            errors=errors,
+        )
+
+    async def async_step_mfa(self, user_input=None):
+        """Handle the step of mfa validation."""
+        errors = {}
+
+        auth_module = self._auth_manager.get_auth_mfa_module(
+            self._auth_module_id)
+        if auth_module is None:
+            # Given an invalid input to async_ste[_select_mfa_module
+            # will show invalid_auth_module error
+            return await self.async_step_select_mfa_module(
+                user_input={'multi_factor_auth_module': None})
+
+        if user_input is not None:
+            expires = self.created_at + SESSION_EXPIRATION
+            if dt_util.utcnow() > expires:
+                errors['base'] = 'login_expired'
+            else:
+                result = await auth_module.async_validation(
+                    self._user.id, user_input)
+                if not result:
+                    errors['base'] = 'invalid_auth'
+
+            if not errors:
+                return await self.async_finish(self._username, mfa_valid=True)
+
+        return self.async_show_form(
+            step_id='mfa',
+            data_schema=auth_module.input_schema,
+            errors=errors,
+        )
+
+    async def async_finish(self, username, mfa_valid=False):
+        """Handle the pass of login flow."""
+        if not mfa_valid:
+            self._username = username
+
+            # We need get user from username, so we need get credential first
+            # async_get_or_create_credentials would not save data, it is safe
+            # to be called here. It will be called later in end of workflow.
+            credentials = await self._auth_provider.\
```

Comment:
Wait what's going on here?

Credentials are an opaque data format. Yes, our Home Assistant provider uses `{'username': username}` but that's just one auth provider. Credentials could also be a user ID in some external system etc.

`async_get_or_create_credentials` takes in the flow result.